File: Habitica/src/test/java/com/habitrpg/android/habitica/utils/DateDeserializerTest.java
package com.habitrpg.android.habitica.utils;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;

import org.junit.Before;
import org.junit.Test;

import java.lang.reflect.Type;
import java.util.Date;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;

public class DateDeserializerTest {
    DateDeserializer deserializer;
    JsonDeserializationContext deserializationContext;
    JsonSerializationContext serializationContext;

    Long referenceTimestamp;

    @Before
    public void setup() {
        this.deserializer = new DateDeserializer();
        this.deserializationContext = new JsonDeserializationContext() {
            @Override
            public <T> T deserialize(JsonElement json, Type typeOfT) throws JsonParseException {
                return null;
            }
        };
        this.serializationContext = new JsonSerializationContext() {
            @Override
            public JsonElement serialize(Object src) {
                return null;
            }

            @Override
            public JsonElement serialize(Object src, Type typeOfSrc) {
                return null;
            }
        };

        this.referenceTimestamp = Long.valueOf("1443445200000");
    }

    @Test
    public void validateNormalDateDeserialize() {
        Date date = this.deserializer.deserialize(new JsonPrimitive("2015-09-28T13:00:00.000Z"), Date.class, this.deserializationContext);

        assertThat(date, is(new Date(referenceTimestamp)));
    }

    @Test
    public void validateTimestampDeserialize() {
        Date date = this.deserializer.deserialize(new JsonPrimitive(referenceTimestamp), Date.class, this.deserializationContext);
        assertThat(date, is(new Date(referenceTimestamp)));
    }

    @Test
    public void validateEmptyDeserialize() {
        Date date = this.deserializer.deserialize(new JsonPrimitive(""), Date.class, this.deserializationContext);
        assertNull(date);
    }

    @Test
    public void validateNormalDateSerialize() {
        JsonElement dateElement = this.deserializer.serialize(new Date(referenceTimestamp), Date.class, this.serializationContext);
        assertThat(dateElement.getAsString(), is("2015-09-28T13:00:00.000Z"));
    }

    @Test
    public void validateEmptySerialize() {
        JsonElement dateElement = this.deserializer.serialize(null, Date.class, this.serializationContext);
        assertThat(dateElement.getAsString(), is(""));
    }
}

File: Habitica/src/test/java/com/habitrpg/android/habitica/utils/DateDeserializerTest.kt
package com.habitrpg.android.habitica.utils

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import io.kotest.matchers.shouldBe
import java.util.Date

class DateDeserializerTest :
    SerializerSpec({
        val deserializer = DateDeserializer()
        val referenceTimestamp = 1443445200000

        "deserialize" should {
            "validate normal date" {
                val date =
                    deserializer.deserialize(
                        JsonPrimitive("2015-09-28T13:00:00.000Z"),
                        Date::class.java,
                        deserializationContext,
                    )
                date shouldBe Date(referenceTimestamp)
            }

            "validate timestamp" {
                val date =
                    deserializer.deserialize(
                        JsonPrimitive(referenceTimestamp),
                        Date::class.java,
                        deserializationContext,
                    )
                date shouldBe Date(referenceTimestamp)
            }

            "validate empty" {
                val date =
                    deserializer.deserialize(JsonPrimitive(""), Date::class.java, deserializationContext)
                date shouldBe null
            }

            "use the first entry of an array" {
                val array = JsonArray().apply { add("2015-09-28T13:00:00.000Z") }
                deserializer.deserialize(array, Date::class.java, deserializationContext) shouldBe Date(referenceTimestamp)
            }

            "return null for an empty array" {
                deserializer.deserialize(JsonArray(), Date::class.java, deserializationContext) shouldBe null
            }

            "parse dates without milliseconds or time" {
                deserializer.deserialize(JsonPrimitive("2015-09-28T13:00:00Z"), Date::class.java, deserializationContext) shouldBe
                    Date(referenceTimestamp)
                deserializer.deserialize(JsonPrimitive("2015-09-28"), Date::class.java, deserializationContext) shouldBe
                    Date(1443398400000)
            }

            "return null for unparseable strings" {
                deserializer.deserialize(JsonPrimitive("not a date"), Date::class.java, deserializationContext) shouldBe null
            }
        }

        "serialize" should {
            "serialize normal date" {
                val dateElement: JsonElement =
                    deserializer.serialize(
                        Date(
                            referenceTimestamp,
                        ),
                        Date::class.java,
                        serializationContext,
                    )
                dateElement.asString shouldBe "2015-09-28T13:00:00.000Z"
            }

            "serialize empty" {
                val dateElement: JsonElement =
                    deserializer.serialize(null, Date::class.java, serializationContext)
                dateElement.asString shouldBe ""
            }
        }
    })
